torch/csrc/jit/passes/autocast.cpp

```diff
@@ -81,21 +81,43 @@ c10::optional<AutocastScope> parseAutocast(Value* value) {
 void castTensorInputs(Node* node, Symbol cast_op) {
   const auto graph = node->owningGraph();
 
-  WithInsertPoint insert_point(node);
-
   std::unordered_set<Value*> casted_inputs;
   for (auto input : node->inputs()) {
     if (input->type()->kind() == TensorType::Kind) {
       casted_inputs.insert(input);
     }
   }
 
+  WithInsertPoint insert_point(node);
+
   for (auto input : casted_inputs) {
     const auto new_input = graph->insert(cast_op, {input});
     node->replaceInputWith(input, new_input);
   }
 }
 
+bool hasExplicitDtypeArgument(Node* node) {
+  const auto& actual_args = node->inputs();
+  const auto& formal_args = node->schema().arguments();
+  TORCH_INTERNAL_ASSERT(actual_args.size() == formal_args.size());
+
+  // Try to identify the `dtype` optional paramater
+  Value* dtype_arg = nullptr;
+  for (size_t i = 0; i < formal_args.size(); ++i) {
+    const auto& formal = formal_args[i];
+    if (auto type = formal.type()->cast<OptionalType>()) {
```

Comment:
Will this work even if formal.type() isn't actually an OptionalType?

Reply:
Can you please elaborate? I'm not sure I understand the question.

Reply:
Is it always legal to cast the `type()` pointer to an `OptionalType`? Maybe it is, I don't know how they're defined.

Reply:
I see, thanks for clarifying the question. That's why we have the `if` there - cast<T> would return nullptr if the cast is not valid.